### PR TITLE
logging improvements

### PR DESCRIPTION
**Pull request by hstojic**
2 main things:
1. now logging initial model fitting as step 0 
2. allow dataset as an argument to log method to allow for data-based model diagnostics

few smaller things
- updated notebook a bit to reflect the changes
- also removed restriction that step numer is non-negative, its not really necessary and in ask-tell case users might want to log steps differently

### CODE REVIEW

File: trieste/bayesian_optimizer.py [hunk 1]
```diff
@@ -669,11 +669,23 @@ def optimize(
...
-                        if step == 0 and fit_initial_model:
+                        if step == 1 and fit_initial_model:
                             for tag, model in models.items():
                                 dataset = datasets[tag]
                                 model.update(dataset)
                                 model.optimize(dataset)
+                            
+                    if step == 1 and fit_initial_model and summary_writer:
+                        logging.set_step_number(0)
+                        with summary_writer.as_default(step=0):
```
Comment (hstojic) [inline trieste/bayesian_optimizer.py:678]: @uri-granta for some reason this didn't work without first doing `logging.set_step_number(0)`, although if I understand correctly it should have worked without issues...

Comment (uri-granta) [reply trieste/bayesian_optimizer.py:678]: The summary writer step argument is separate from the logging step state so you have to specify both. You can use a context manager if you prefer though, so you don't have to worry about resetting it afterwards.

```suggestion
                        with logging.step_number(0):
                            with summary_writer.as_default(step=0):
```

Comment (hstojic) [reply trieste/bayesian_optimizer.py:678]: thanks for the suggestion, that's definitely a better way to do it

Comment (hstojic) [reply trieste/bayesian_optimizer.py:678]: ok, that didnt work at the end for logging.scalar line below, logging wallclock model fitting time, so I reverted back

File: trieste/bayesian_optimizer.py [hunk 2]
```diff
@@ -667,14 +667,25 @@ def optimize(
                             "will be available."
                         ) from e
 
-                with Timer() as total_step_wallclock_timer:
```
Comment (hstojic) [inline trieste/bayesian_optimizer.py:670]: @uri-granta I have removed this bit as initial model fitting wallclock is now shown as step 0, so also excluded from total step time

File: trieste/bayesian_optimizer.py [hunk 3]
```diff
@@ -707,7 +718,6 @@ def optimize(
                             models,
                             query_points,
                             tagged_output,
-                            initial_model_fitting_timer,
```
Comment (hstojic) [inline trieste/bayesian_optimizer.py:710]: not needed anymore as logged separately now

File: trieste/logging.py [hunk 4]
```diff
@@ -83,8 +83,6 @@ def set_step_number(step_number: int) -> None:
     :raise ValueError: if step_number < 0
     """
     global _STEP_NUMBER
-    if step_number < 0:
-        raise ValueError(f"step_number must be non-negative (got {step_number})")
```
Comment (hstojic) [inline trieste/logging.py:87]: unnecessary in my opiniopn, in ask-tell one might log initial model stuff as -1 for example

File: trieste/models/interfaces.py [hunk 5]
```diff
@@ -90,9 +90,11 @@ def predict_y(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
             f"Model {self!r} does not support predicting observations, just the latent function"
         )
 
-    def log(self) -> None:
+    def log(self, dataset: Dataset) -> None:
         """
         Log model-specific information at a given optimization step.
+
+        :param dataset: The data that can be used to log additional data-based model summaries.
```
Comment (hstojic) [inline trieste/models/interfaces.py]: @uri-granta added dataset argument as its more flexible, there are some metrics we would like to log that are based on the training data

Comment (hstojic) [reply trieste/models/interfaces.py]: perhaps we should make it optional though?

File: docs/notebooks/visualizing_with_tensorboard.pct.py [hunk 6]
```diff
@@ -153,17 +153,18 @@ def acquire(self, search_space, models, datasets=None):
...
 # ## Using Tensorboard with Ask-Tell Optimization
 #
-# To use Tensorboard logging with the [Ask-Tell interface](ask_tell_optimization.ipynb), you must also explicitly set the optimization step number before calling ask or tell:
+# To use Tensorboard logging with the [Ask-Tell interface](ask_tell_optimization.ipynb), you must also explicitly set the optimization step number before calling ask or tell methods. If fitting the model on initialization of the Ask-Tell interface (default), then you must also set the optimization step number before initializing the interface to log initial model fitting as a different step.
 
 # %%
 summary_writer = tf.summary.create_file_writer("logs/tensorboard/experiment4")
 trieste.logging.set_tensorboard_writer(summary_writer)
 
+trieste.logging.set_step_number(0)
```
Comment (uri-granta) [inline docs/notebooks/visualizing_with_tensorboard.pct.py:162]: (step number does deault to 0, so strictly speaking this isn't necessary, but probably best to leave it in)

File: trieste/logging.py (hunk 4 above)
Comment (uri-granta) [reply trieste/logging.py:87]: I don't think that works with tensorboard (can't remember if you get an error or if the data just disappears)

File: trieste/models/gpflow/interface.py [hunk 7]
```diff
@@ -144,9 +144,11 @@ def optimize(self, dataset: Dataset) -> None:
         """
         self.optimizer.optimize(self.model, dataset)
 
-    def log(self) -> None:
+    def log(self, dataset: Dataset) -> None:
         """
         Log model-specific information at a given optimization step.
+
+        :param dataset: The data that can be used to log additional data-based model summaries.
```
Comment (uri-granta) [inline trieste/models/gpflow/interface.py]: Should probably be optional and default to None (like acquisition function builders)

File: trieste/models/interfaces.py (hunk 5 above)
Comment (uri-granta) [reply trieste/models/interfaces.py]: Yes, make this optional, and default it to None (like acquisition function builders)

File: trieste/bayesian_optimizer.py (hunk 3 above)
Comment (uri-granta) [reply trieste/bayesian_optimizer.py:710]: :+1:

File: trieste/bayesian_optimizer.py (hunk 1 above)
Comment (uri-granta) [reply trieste/bayesian_optimizer.py:678]: Was the indentation correct? What error did you get? The context manager should be totally equivalent to setting and resetting like you're doing now, but cleaner and more robust.

File: trieste/bayesian_optimizer.py [hunk 8]
```diff
@@ -667,14 +667,25 @@ def optimize(
...
+                            model.update(dataset)
+                            model.optimize(dataset)
+                    if summary_writer:
+                        logging.set_step_number(0)
+                        with summary_writer.as_default(step=0):
                             for tag, model in models.items():
-                                dataset = datasets[tag]
-                                model.update(dataset)
-                                model.optimize(dataset)
+                                with tf.name_scope(f"{tag}.model"):
```
Comment (uri-granta) [inline trieste/bayesian_optimizer.py]: maybe move this to a `_write_summary_initial_fit` function (to go alongside `_write_summary_step` and `_write_summary_init`)?

File: tests/unit/test_logging.py [hunk 9]
```diff
@@ -88,11 +88,6 @@ def test_set_get_step_number(step: int) -> None:
     assert get_step_number() == 0
 
 
-def test_set_step_number_error() -> None:
-    with pytest.raises(ValueError):
-        set_step_number(-1)
```
Comment (uri-granta) [inline tests/unit/test_logging.py:93]: (remember to restore this if you bring back the check)

File: tests/unit/models/gpflow/test_models.py [hunk 10]
```diff
@@ -263,14 +263,15 @@ def test_gpflow_wrappers_log(
...
+    data = Dataset(x, fnc_3x_plus_10(x))
 
     if kernel is not None:
         model.model.kernel = kernel
 
     mocked_summary_writer = unittest.mock.MagicMock()
     with tensorboard_writer(mocked_summary_writer):
         with step_number(42):
-            model.log()
+            model.log(data)
```
Comment (uri-granta) [inline tests/unit/models/gpflow/test_models.py]: (If you make data optional you can skip passing it in here)

**Review by uri-granta — COMMENTED** [review]
Looks good. Various comments.

File: trieste/bayesian_optimizer.py (hunk 1 above)
Comment (hstojic) [reply trieste/bayesian_optimizer.py:678]: yes, indentation was correct I think, there was no error, scalar simply wasn't logged at all

File: trieste/logging.py (hunk 4 above)
Comment (hstojic) [reply trieste/logging.py:87]: it works, tried it with ask-tell interface

File: trieste/models/interfaces.py (hunk 5 above)
Comment (hstojic) [reply trieste/models/interfaces.py]: done!

File: trieste/bayesian_optimizer.py (hunk 8 above)
Comment (hstojic) [reply trieste/bayesian_optimizer.py]: done

**Review by uri-granta — APPROVED** [review]
LGTM!

**Pull request merged**